### PR TITLE
kv: increase dist sender's RPC timeout

### PR DESCRIPTION
**Pull request by tamird**

### CODE REVIEW

**Review by spencerkimball — APPROVED** [review]
Yes, I'm not too worried about this change. This is a relatively rare occurrence and I'm not convinced there's going to be a big win to allow these ambiguous results to return after a wait. In those cases, it's likely that the live node (live because gRPC heartbeats are still working) is blocked up in a Raft group or similar. In those cases, there are all kinds of other problems, so letting this pending request continue waiting isn't likely to make the system more responsive.

**Comment by andreimatei** [comment]
FWIW, I think what we want for the DistSender timeouts is some mechanism to differentiate between slow RPCs that have been accepted by the lease holder versus RPCs that haven't even been accepted yet (and thus we don't really know if the recipient even is the leaseholder). In lack of that, I always disliked those timeouts, so I think this change is great.

**Review by tbg — APPROVED** [review]

**Comment by bdarnell** [comment]
This isn't the timeout you want to remove - This controls how long we wait when we already have multiple RPCs in flight, so removing it means we'll wait as long as it takes for them to complete. You want to remove SendNextTimeout to avoid having multiple RPCs in flight in the first place.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16088#-:-Kkv-NL7Ac2-8tjxjPs3:b-pd4sza)*

**Comment by tschottdorf** [comment]
@bdarnell I also chatted with @tamird because I thought the PR description suggested that's what he wanted to do, but it seems intentional that he's only removing this smaller bit now (which to me seems also reasonable, and perhaps less contentious). Perhaps the PR message should be clarified though.

**Comment by bdarnell** [comment]
Ah. Well in that case I disagree with this change. The reliance on a single leader means that we may not need to send RPCs to multiple replicas, but having sent them, they may be slow and we should use timeouts when waiting on secondary RPC attempts.

File: pkg/kv/dist_sender.go [hunk 1]
```diff
@@ -1230,26 +1229,19 @@ func (ds *DistSender) sendToReplicas(
...
-						select {
-						case pendingCall := <-done:
-							pending--
-							if err := pendingCall.Err; err != nil {
-								if grpc.Code(err) != codes.Unavailable {
-									ambiguousResult = true
-								}
-							} else if pendingCall.Reply.Error == nil {
-								return pendingCall.Reply, nil
+					for ; pending > 0; pending-- {
```
Comment (spencerkimball) [inline pkg/kv/dist_sender.go]: There's actually no point in waiting for these pending RPCs unless `haveCommit` is true. Let's change this line to:

`for ; haveCommit && pending > 0; pending-- {`

**Comment by spencerkimball** [comment]
I think @bdarnell has a point: removing this timeout entirely means a client can become permanently stuck if a range is jammed and not processing Raft commands. I think it might make more sense to increase the value of the timeout instead (perhaps to `10s`?). @tamird's main concern is that this is breaking benchmarking because we do end up with sloooooow RPCs in some cases.

We can rely on the gRPC heartbeat timeout to cover the case of a connection which has become too unhealthy to serve RPCs. That's a `3s` timeout. Let's leave this code in place and up the timeout to further attenuate occurrences of `AmbiguousResultError` while also protecting clients from permanent "black hole" ranges.

Also, @tamird, please note my comment around `haveCommit` being used to avoid waiting at all for pending RPCs.

**Comment by petermattis** [comment]
This could only break multi-node benchmarks, right? If this is somehow affecting single-node benchmarks I think there is a bug somewhere.

File: pkg/kv/dist_sender.go (hunk 1 above)
Comment (tamird) [inline pkg/kv/dist_sender.go]: Done.

**Comment by tamird** [comment]
Revised this to increase the timeout instead of removing it, PTAL

**Comment by tamird** [comment]
@petermattis indeed, only affects multi-node benchmarks.

**Comment by petermattis** [comment]
@tamird The multi-node `sql` benchmarks are somewhat uninteresting in that the nodes are all running on the same machine. That's not to dissuade this PR, but to help you unblock the benchmark storage work. I'd be fine in the short term with only recording the single-node `sql` benchmarks.

**Comment by tamird** [comment]
Yep, I'm proceeding that way.

**Comment by spencerkimball** [comment]
lgtm

**Comment by bdarnell** [comment]
:lgtm:

---

Reviewed 6 of 7 files at r3, 1 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16088#-:-Kkvh_S-sN8RUNdkKtkN:bnfp4nl)*

**Pull request merged**